// incyclist-ant-plus/lib/sensors/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SpeedCadenceSensorState = exports.SpeedCadenceSensor = exports.SpeedSensorState = exports.SpeedSensor = exports.CadenceSensorState = exports.CadenceSensor = exports.FitnessEquipmentSensorState = exports.FitnessEquipmentSensor = exports.HeartRateSensorState = exports.HeartRateSensor = exports.BicyclePowerSensorState = exports.BicyclePowerSensor = void 0;
const bicycle_power_sensors_1 = __importStar(require("./bicycle-power-sensors"));
exports.BicyclePowerSensor = bicycle_power_sensors_1.default;
Object.defineProperty(exports, "BicyclePowerSensorState", { enumerable: true, get: function () { return bicycle_power_sensors_1.BicyclePowerSensorState; } });
const heart_rate_sensors_1 = __importStar(require("./heart-rate-sensors"));
exports.HeartRateSensor = heart_rate_sensors_1.default;
Object.defineProperty(exports, "HeartRateSensorState", { enumerable: true, get: function () { return heart_rate_sensors_1.HeartRateSensorState; } });
const fitness_equipment_sensors_1 = __importStar(require("./fitness-equipment-sensors"));
exports.FitnessEquipmentSensor = fitness_equipment_sensors_1.default;
Object.defineProperty(exports, "FitnessEquipmentSensorState", { enumerable: true, get: function () { return fitness_equipment_sensors_1.FitnessEquipmentSensorState; } });
const cadence_sensors_1 = __importStar(require("./cadence-sensors"));
exports.CadenceSensor = cadence_sensors_1.default;
Object.defineProperty(exports, "CadenceSensorState", { enumerable: true, get: function () { return cadence_sensors_1.CadenceSensorState; } });
const speed_sensors_1 = __importStar(require("./speed-sensors"));
exports.SpeedSensor = speed_sensors_1.default;
Object.defineProperty(exports, "SpeedSensorState", { enumerable: true, get: function () { return speed_sensors_1.SpeedSensorState; } });
const speed_cadence_sensors_1 = __importStar(require("./speed-cadence-sensors"));
exports.SpeedCadenceSensor = speed_cadence_sensors_1.default;
Object.defineProperty(exports, "SpeedCadenceSensorState", { enumerable: true, get: function () { return speed_cadence_sensors_1.SpeedCadenceSensorState; } });
