// incyclist-ant-plus/lib/ant.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AntPlusScanner = exports.AntPlusSensor = exports.GarminStick3 = exports.GarminStick2 = exports.Constants = exports.Messages = void 0;
const base_sensor_1 = __importDefault(require("./base-sensor"));
const stick_1 = __importDefault(require("./stick"));
var messages_1 = require("./messages");
Object.defineProperty(exports, "Messages", { enumerable: true, get: function () { return messages_1.Messages; } });
var consts_1 = require("./consts");
Object.defineProperty(exports, "Constants", { enumerable: true, get: function () { return consts_1.Constants; } });
__exportStar(require("./ant-device"), exports);
class GarminStick2 extends stick_1.default {
    constructor(props = {}) {
        super(0x0fcf, 0x1008, props);
    }
}
exports.GarminStick2 = GarminStick2;
class GarminStick3 extends stick_1.default {
    constructor(props = {}) {
        super(0x0fcf, 0x1009, props);
    }
}
exports.GarminStick3 = GarminStick3;
class AntPlusSensor extends base_sensor_1.default {
    attach(channel, type, deviceID, deviceType, transmissionType, timeout, period) {
        return super.attach(channel, type, deviceID, deviceType, transmissionType, timeout, period, 57);
    }
}
exports.AntPlusSensor = AntPlusSensor;
class AntPlusScanner extends base_sensor_1.default {
    scan(type) {
        return super.scan(type, 57);
    }
}
exports.AntPlusScanner = AntPlusScanner;
